[IRBuilder] Fold binary intrinsics

agentcooper opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/61240.

agentcooper committed:
[IRBuilder] Fold binary intrinsics

Fixes https://github.com/llvm/llvm-project/issues/61240.

nikic reviewed (commented):
The high level approach here should be to not hard-code constant folding for specific intrinsics, but rather make use of generic APIs like ConstantFoldCall or simplifyCall.

This is not entirely straightforward because these APIs currently require an already existing call, so some refactoring would be necessary to allow calling them with e.g. just a function, function type and arguments or something like that.

agentcooper committed:
Reuse simplifyBinaryIntrinsic

agentcooper commented:
@nikic Thanks for the guidance!

I've just pushed a commit where I am able to reuse `simplifyBinaryIntrinsic` in case of the `InstSimplifyFolder`. I am still not entirely sure what should I do in the `ConstantFolder` – the `ConstantFoldCall` you've mentioned is a part of Analysis and the ConstantFolder is in the IR.

agentcooper committed:
Merge branch 'llvm:main' into fold-binary-intrinsics

agentcooper committed:
Change return type of CreateBinaryIntrinsic from CallInst to Value

agentcooper committed:
Code formatting

agentcooper commented:
I've implemented `FoldBinaryIntrinsic` in all 3 folders: ConstantFolder, TargetFolder, InstSimplifyFolder.

In case of the TargetFolder I split the `ConstantFoldScalarCall2` into 2 functions: `ConstantFoldLibCall2` that handles `LibFunc` and `ConstantFoldIntrinsicCall2` that handles intrinsics and is reused.

I've also switched the return type for `CreateBinaryIntrinsic` from `CallInst` to `Value` to reflect the fact that it can be folded.

In case of ConstantFolder I've created a new `ConstantFoldBinaryIntrinsicInstruction` which now has the code that was previously in SLPVectorizer.

Not sure why Windows build is failing, but Linux one is passing.

@nikic Could you please take another look?

agentcooper commented:
@nikic ping :-)

dtcxzyw commented on llvm/include/llvm/Analysis/ConstantFolding.h:
```diff
@@ -22,6 +22,11 @@
 namespace llvm {
+
+namespace Intrinsic {
+typedef unsigned ID;
```
```suggestion
using ID = unsigned;
```

dtcxzyw commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -767,19 +767,25 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+      if (auto *CI = dyn_cast<CallInst>(V)) {
+        CI->copyFastMathFlags(&II);
+        CI->takeName(&II);
+        return IC.replaceInstUsesWith(II, CI);
```
```suggestion
```

dtcxzyw commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -767,19 +767,25 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
     if (match(Src0, PatternMatch::m_NaN()) || isa<UndefValue>(Src0)) {
-      NewCall = IC.Builder.CreateMinNum(Src1, Src2);
+      V = IC.Builder.CreateMinNum(Src1, Src2);
+      // llvm::dbgs() << "1 " << *V << "\n";
```
Please remove these debug printing statements.

dtcxzyw commented on llvm/lib/IR/ConstantFold.cpp:
```diff
@@ -1735,3 +1735,29 @@ Constant *llvm::ConstantFoldGetElementPtr(Type *PointeeTy, Constant *C,
+    break;
+  }
+  return nullptr;
+}
```
No newline at the end of the file
```suggestion
}

```

dtcxzyw commented on llvm/include/llvm/Analysis/InstSimplifyFolder.h:
```diff
@@ -117,6 +117,11 @@ class InstSimplifyFolder final : public IRBuilderFolder {
 
+  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS,
+                             Value *RHS) const override {
+    return simplifyBinaryIntrinsic(ID, LHS->getType(), LHS, RHS, SQ, nullptr);
```
`LHS->getType()` is incorrect for `@llvm.is.fpclass`. I don't know why `abs/ctlz/cttz/is.fpclass` are treated as binary intrinsics.

agentcooper replied:
I've updated the `FoldBinaryIntrinsic` definition to include a `Type *` parameter, which is be passed from `CreateBinaryIntrinsic` using `F->getReturnType()`.

dtcxzyw commented on llvm/lib/IR/IRBuilder.cpp:
```diff
@@ -918,10 +918,11 @@ CallInst *IRBuilderBase::CreateUnaryIntrinsic(Intrinsic::ID ID, Value *V,
+Value *IRBuilderBase::CreateBinaryIntrinsic(Intrinsic::ID ID, Value *LHS,
+                                            Value *RHS, Instruction *FMFSource,
+                                            const Twine &Name) {
+  if (auto *V = Folder.FoldBinaryIntrinsic(ID, LHS, RHS))
```
Can we pass the argument `FMFSource` into `Folder.FoldBinaryIntrinsic`?

agentcooper replied:
It is not immediately clear to me how it can be used, given that `FoldBinaryIntrinsic` returns a `Constant` and from my understanding the `FMFSource` is used to copy FMF flags when a new `CallInst` is created.

dtcxzyw replied:
We can use FMF flags to further fold expressions in `InstSimplifyFolder`.
E.g., `minimum(X, -inf) -> -inf when FMF nnan is set`.

agentcooper committed:
Add type parameter to FoldBinaryIntrinsic

RKSimon commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -2257,13 +2257,14 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
           llvm_unreachable("unexpected intrinsic ID");
         }
-        Instruction *NewCall = Builder.CreateBinaryIntrinsic(
+        auto *V = Builder.CreateBinaryIntrinsic(
```
Value *

agentcooper committed:
Explicit type annotation

agentcooper committed:
Merge intrinsic cases in SLPVectorizer

agentcooper commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -14116,28 +14116,12 @@ class HorizontalReduction {
-        return ConstantFP::get(LHS->getType(),
-                               minimum(cast<ConstantFP>(LHS)->getValueAPF(),
-                                      cast<ConstantFP>(RHS)->getValueAPF()));
       return Builder.CreateBinaryIntrinsic(Intrinsic::minimum, LHS, RHS);
```
Nice! Done.

agentcooper committed:
Pass FMFSource

dtcxzyw commented on llvm/include/llvm/Analysis/InstSimplifyFolder.h:
```diff
@@ -117,9 +117,10 @@ class InstSimplifyFolder final : public IRBuilderFolder {
-                             Type *Ty) const override {
-    return simplifyBinaryIntrinsic(ID, Ty, LHS, RHS, SQ, nullptr);
+  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS, Type *Ty,
+                             Instruction *FMFSource = nullptr) const override {
```
```suggestion
                             Instruction *FMFSource) const override {
```
Don't use default argument in an overridden virtual function.

dtcxzyw commented on llvm/include/llvm/Analysis/TargetFolder.h:
```diff
@@ -191,12 +191,12 @@ class TargetFolder final : public IRBuilderFolder {
-  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS,
-                             Type *Ty) const override {
+  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS, Type *Ty,
+                             Instruction *FMFSource = nullptr) const override {
```
```suggestion
                             Instruction *FMFSource) const override {
```

dtcxzyw commented on llvm/include/llvm/Analysis/TargetFolder.h:
```diff
@@ -191,12 +191,12 @@ class TargetFolder final : public IRBuilderFolder {
     auto *C2 = dyn_cast<Constant>(RHS);
     if (C1 && C2)
-      return ConstantFoldBinaryIntrinsic(ID, C1, C2, Ty);
+      return ConstantFoldBinaryIntrinsic(ID, C1, C2, Ty, FMFSource);
```
It doesn't help constant folding :(

dtcxzyw replied:
Oh, my bad. It is used to fold constrained fp intrinsics.

dtcxzyw commented on llvm/include/llvm/IR/ConstantFolder.h:
```diff
@@ -183,8 +183,8 @@ class ConstantFolder final : public IRBuilderFolder {
-  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS,
-                             Type *Ty) const override {
+  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS, Type *Ty,
+                             Instruction *FMFSource = nullptr) const override {
```
```suggestion
                             Instruction *FMFSource) const override {
```

dtcxzyw commented on llvm/include/llvm/IR/NoFolder.h:
```diff
@@ -112,8 +112,8 @@ class NoFolder final : public IRBuilderFolder {
-  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS,
-                             Type *Ty) const override {
+  Value *FoldBinaryIntrinsic(Intrinsic::ID ID, Value *LHS, Value *RHS, Type *Ty,
+                             Instruction *FMFSource = nullptr) const override {
```
```suggestion
                             Instruction *FMFSource) const override {
```

agentcooper committed:
Remove default argument in overridden virtual functions

dtcxzyw reviewed (approved):
LGTM. Thanks!
Please wait for additional approval from other reviewers.

nikic commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -767,19 +767,21 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
-      NewCall->takeName(&II);
-      return IC.replaceInstUsesWith(II, NewCall);
+    if (V) {
+      if (auto *CI = dyn_cast<CallInst>(V)) {
```
Why? dyn_cast should always be used with auto.

nikic replied:
For reference, the guideline for this is at https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable.

nikic reviewed (changes requested):
Please remove the changes to IR/ConstantFold. I'm not willing to duplicate constant folding logic in there. If we *really* wanted to we could *move* logic there, but we should not copy it.

agentcooper commented:
@nikic So the suggestion is to keep the SLPVectorizer code as it is right now and maybe create a separate PR to move SLPVectorizer to TargetFolder instead? Wouldn't you consider it an improvement to extract the existing folding code from SLPVectorizer into IR/ConstantFolder?

RKSimon commented:
@agentcooper reverse-ping?

nikic commented:
> @nikic So the suggestion is to keep the SLPVectorizer code as it is right now and maybe create a separate PR to move SLPVectorizer to TargetFolder instead?

Yes, exactly! (Or move SLPVectorizer to TargetFolder first.)

> Wouldn't you consider it an improvement to extract the existing folding code from SLPVectorizer into IR/ConstantFolder?

No, because it sets a bad precedent. The ConstantFolding / ConstantFold split is already annoying enough as it is, without introducing actual duplication between them.

agentcooper committed:
Remove changes to IR/ConstantFold and SLPVectorizer

agentcooper committed:
Feedback

agentcooper commented:
@nikic Got it. I updated the PR to exclude changes to IR/ConstantFold and SLPVectorizer (FYI @RKSimon).

agentcooper commented:
@nikic ping :-)

nikic reviewed (approved):
LGTM

agentcooper commented:
@nikic @dtcxzyw If this has enough reviews, could you please merge it for me? I don't have commit access.
